services/user.go

```diff
@@ -333,3 +383,432 @@ func (srv *UserService) checkUpdateCascade() bool {
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to begin registration: %w", err)
+	}
+
+	// Store session data in cache instead of database to avoid serialization issues
+	sessionBytes, err := json.Marshal(sessionData)
+	if err != nil {
+		return nil, nil, fmt.Errorf(errFailedMarshalSessionData, err)
+	}
+
+	// Use cache with 5 minute expiry for session data
+	cacheKey := fmt.Sprintf(webAuthnSessionCacheKeyFormat, user.ID)
+	srv.cache.Set(cacheKey, string(sessionBytes), 5*time.Minute)
+
+	// Store credential name separately in cache for use during finish registration
+	credentialNameKey := fmt.Sprintf(webAuthnCredentialNameCacheFormat, user.ID)
+	if credentialName != "" {
+		srv.cache.Set(credentialNameKey, credentialName, 5*time.Minute)
+	} else {
+		// Use default name if none provided
+		defaultName := fmt.Sprintf("Security Key %d", len(user.GetWebAuthnCredentials())+1)
+		srv.cache.Set(credentialNameKey, defaultName, 5*time.Minute)
+	}
+
+	return options, string(sessionBytes), nil
+}
+
+func (srv *UserService) WebAuthnFinishRegistration(user *models.User, sessionData interface{}, credentialCreationResponse interface{}) error {
+	if srv.webAuthn == nil {
+		return errors.New(errWebAuthnNotEnabled)
+	}
+
+	// Check if sessionData is nil
+	if sessionData == nil {
+		return errors.New("session data is missing")
+	}
+
+	// Parse session data
+	sessionDataStruct := &webauthn.SessionData{}
+	sessionDataStr, ok := sessionData.(string)
+	if !ok {
+		return errors.New("session data must be a string")
+	}
+
+	if err := json.Unmarshal([]byte(sessionDataStr), sessionDataStruct); err != nil {
+		return fmt.Errorf(errFailedUnmarshalSessionData, err)
+	}
+
+	// Verify session exists in cache
+	cacheKey := fmt.Sprintf(webAuthnSessionCacheKeyFormat, user.ID)
+	if _, found := srv.cache.Get(cacheKey); !found {
+		return errors.New(errSessionExpiredOrInvalid)
+	}
+
+	// Parse credential creation response
+	ccr, err := protocol.ParseCredentialCreationResponseBody(bytes.NewReader(credentialCreationResponse.([]byte)))
+	if err != nil {
+		return fmt.Errorf("failed to parse credential creation response: %w", err)
+	}
+
+	credential, err := srv.webAuthn.CreateCredential(user, *sessionDataStruct, ccr)
+	if err != nil {
+		return fmt.Errorf("failed to create credential: %w", err)
+	}
+
+	// Retrieve credential name from cache
+	credentialNameKey := fmt.Sprintf(webAuthnCredentialNameCacheFormat, user.ID)
+	credentialName := "Security Key"
+	if cachedName, found := srv.cache.Get(credentialNameKey); found {
+		if nameStr, ok := cachedName.(string); ok {
+			credentialName = nameStr
+		}
+	}
+
+	// Convert to our model and add to user
+	webAuthnCred := &models.WebAuthnCredential{
+		ID:              fmt.Sprintf("cred_%d", time.Now().Unix()),
+		Name:            credentialName,
+		CredentialID:    credential.ID,
+		PublicKey:       credential.PublicKey,
+		AttestationType: credential.AttestationType,
+		Transport:       credential.Transport,
+		Flags: struct {
```

Comment:
Please don't use an anonymous struct here to prevent boilerplate code. Same for `Authenticator` below.